Workshop overhaul - Attester section (#285)
KILTprotocol/docs: docs/develop/03_workshop/04_attester/03_ctype.md

<!-- Taken from https://github.com/webpack-contrib/raw-loader/issues/91#issuecomment-648830498 -->
import Ctype from '@site/scripts/out/ctype.json.raw!=!raw-loader!@site/scripts/out/ctype.json';

A claim type (CType) is a KILT-specific term, but the concept is simple:
A CType is a JSON schema that defines the structure of a claim, and you can think of it as the data model for your claim.

:::info CType

A CType ensures that a credential contains all required attributes, e.g., a driver's license has to contain a name, date of birth, and the vehicle types that the claimer can drive.

> The CType is important since a Verifier requests credentials for a specific CType. For example, the traffic police want to see your driver's license, not your gym membership.

Reviewer: Split the sentence to a new line after the dot 😃

```suggestion
The CType is important since a Verifier requests credentials for a specific CType.
For example, the traffic police want to see your driver's license, not your gym membership.
```

To learn more about CTypes, read the [in-depth CType documentation](../../../concepts/05_credentials/02_ctypes.md).
You can also [read through existing CTypes in the CType-index](https://github.com/KILTprotocol/ctype-index).
:::

Before the <span className="label-role attester">Attester</span> can attest credentials, they must decide which CType they support.

> For example, a traffic authority only issues driver's licenses (A CType for driver's license), not trade register excerpts.

Reviewer: Maybe an easier example than a trade register excerpt? Like a diploma?

| `$id`        | The KILT id of this CType. It's the most important property as it represents the **digital footprint** of the CType.                                               |
| `$schema`    | A reference to the meta-schema describing what a CType may look like. There are two versions.                                                              |
| `title`      | The title of the CType.                                                                                                                                             |
| `properties` | The properties that a claim conforming to this CType may have.                                                                                                      |
| `type` | Type is an object for all CTypes.                                                                                                  |
| `additionalProperties` | The default is false. This restricts unwanted properties in a claim.                                                                                                      |

A CType is stored on the KILT blockchain.

> <!-- TODO: What is a Credential Registry's REST API? -->

Author: What is a Credential Registry's REST API?

Reviewer: There would be a service where CTypes can be fetched. I guess it's more correct to call them CType Registry. Also I would scrape the REST API part.

Author: @ntn-x2 Yes, but what is it? Where can I find more about it? What options are available etc etc?

Reviewer: This is the only service I am aware of: https://ctypehub.galaniprojects.de/. You could also use the REST API that [Subscan](https://spiritnet.subscan.io/) provides.

Reviewer: I think @Dudleyneedham, who was the author of that section, might have more background info.

Reviewer 2: Yes, I think we were describing the service mentioned above. We can remove the REST API.

In a real-world situation, a user would retrieve an existing CType from the chain or a repository of CTypes.

> For example, via a Credential Registry's REST API.

Reviewer: I would just remove this.

Reviewer 2: Agreed.

In this tutorial, the <span className="label-role attester">Attester</span> creates and attempts to store a CType on the KILT test blockchain.

## Create CType

> Copy the following that defines a `CType` with a given schema:

Reviewer: ```suggestion
Copy the following snippet that defines a `CType` with a given schema:
```

Reviewer: Or then, based on the phrasing in the next block

```suggestion
Copy the following to define a `CType` with a given schema:
```

<TsJsBlock fileName="attester/ctypeSchema">
  {CtypeSchema}
</TsJsBlock>

## Get CType

Copy the following to create a `CType` on the chain:

<TsJsBlock fileName="attester/generateCtype">
  {GenerateCtype}
</TsJsBlock>

> <!-- TODO: Already exists and code hangs -->

Author: The CType already exists and the code hangs, which seems reasonable, but what can we do about it?

Reviewer: Add some random factor into the CType. For instance, add the attester's DID inside the name property. As long as the DID is freshly generated, the CType (hash) is guaranteed (with enough probability) to be unique as well. We do something similar in our SDK integration tests: https://github.com/KILTprotocol/sdk-js/blob/c4ab492812d19169532a399b57dd1bd013a61570/tests/integration/Ctypes.spec.ts#L42.

> The `ensureStoredCType` function takes the Attester's account, DID, and a callback to sign the function and checks if the CType is already on-chain.

Reviewer: ```suggestion
The `ensureStoredCType` function takes the Attester's account, DID, and a callback to sign the function and checks if the CType is already on chain.
```

It uses the `verifyStored` method to pass the CType to the KILT blockchain and make the check.

> If it does not exist, it stores it on-chain, using the `toChain` method to encode the CType into a unique hash and the `add` method to create a new CType from the given unique hash and associate it with the Attester.

Reviewer: ```suggestion
If it does not exist, it stores it on chain, using the `toChain` method to encode the CType into a unique hash and the `add` method to create a new CType from the given unique hash and associate it with the Attester.
```

<!-- Taken from https://github.com/webpack-contrib/raw-loader/issues/91#issuecomment-648830498 -->
import Ctype from '@site/scripts/out/ctype.json.raw!=!raw-loader!@site/scripts/out/ctype.json';

> A claim type (CType) is a KILT-specific term, but the concept is simple:

Reviewer 2: I believe at some point we will switch this from claim type to credential type.
